meta-client/src/test/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionPuppy.java

```diff
@@ -0,0 +1,87 @@
+package org.apache.helix.metaclient.recipes.leaderelection;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import org.apache.helix.metaclient.MetaClientTestUtil;
+import org.apache.helix.metaclient.api.MetaClientInterface;
+import org.apache.helix.metaclient.exception.MetaClientException;
+import org.apache.helix.metaclient.puppy.AbstractPuppy;
+import org.apache.helix.metaclient.puppy.PuppySpec;
+import org.testng.Assert;
+
+
+public class LeaderElectionPuppy extends AbstractPuppy {
+  String _leaderGroup;
+  String _participant;
+  private final Random _random;
+  LeaderElectionClient _leaderElectionClient;
+
+  public LeaderElectionPuppy(MetaClientInterface<String> metaclient, PuppySpec puppySpec) {
+    super(metaclient, puppySpec);
+    _random = new Random();
+  }
+
+  public LeaderElectionPuppy(LeaderElectionClient leaderElectionClient, PuppySpec puppySpec, String leaderGroup,
+      String participant) {
+    super(leaderElectionClient.getMetaClient(), puppySpec);
+    _leaderElectionClient = leaderElectionClient;
+    _leaderGroup = leaderGroup;
+    _random = new Random();
+    _participant = participant;
+  }
+
+  @Override
+  protected void bark() throws Exception {
```

Comment:
I see in the cleanup() that you catch the exception and handle it. In the logic of bark() itself, I think we should forcefully check cases where the leaderElection API's should fail. Perhaps I didn't read this right but it seems like all verifications are made with "positive" checks. What do you think?

Reply:
Thanks for review!
Right now all exceptions thrown by leader election clients are actually fatal exceptions.Since each puppy has one client and join/leaving pool are executed in sequence, we should not have any actual exception thorwn.